### PR TITLE
PLT-2563 Redo password recovery to use a database table

### PR DESCRIPTION
**Pull request by jwilander**

### CODE REVIEW

File: api/user.go [hunk 1]
```diff
@@ -1621,33 +1622,42 @@ func resetPassword(c *Context, w http.ResponseWriter, r *http.Request) {
...
-			return
+			go func() {
+				if result := <-Srv.Store.PasswordRecovery().Delete(userId); result.Err != nil {
+					l4g.Error("%v", result.Err)
+				}
+			}()
 		}
 	}
 
+	c.LogAuditWithUserId(userId, "attempt")
```
Comment (crspeller) [inline api/user.go]: Shouldn't this be before we check the validity of the code?

Comment (crspeller) [inline api/user.go]: Also there is no corresponding success audit.

Comment (jwilander) [inline api/user.go]: Hmmm, it never was before we checked the hash but yeah it's probably a good idea to do it before

File: api/user.go [hunk 2]
```diff
@@ -1621,33 +1622,42 @@ func resetPassword(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
 	userId := props["user_id"]
-	hash := props["hash"]
-	timeStr := ""
 
-	if !c.IsSystemAdmin() {
-		if len(hash) == 0 {
-			c.SetInvalidParam("resetPassword", "hash")
+	if !c.IsSystemAdmin() || len(userId) == 0 {
```
Comment (crspeller) [inline api/user.go]: Maybe the system admins ability to reset passwords should be a separate route in the admin routes. Then they can share a helper function that actually does the reset. This function is a bit confusing. Normally not important but it is a critical security function.

File: api/user.go (hunk 1 above)
Comment (jwilander) [inline api/user.go]: There is a success audit already there, just not shown in the diff

**Comment by crspeller** [comment]
Maybe the system admins ability to reset passwords should be a separate route in the admin routes. Then they can share a helper function that actually does the reset. This function is a bit confusing. Normally not too important but it is a critical security function.

**Comment by jwilander** [comment]
Yeah, that's a good point. It really should have been a separate route from the beginning. I'll add it now

**Comment by coreyhulen** [comment]
+1

**Comment by coreyhulen** [comment]
Actually I think we need to perm delete the rows when a user is perm deleted?

**Comment by jwilander** [comment]
Right, I'll make that change too

**Comment by jwilander** [comment]
Ready for re-review @crspeller @coreyhulen

File: api/admin.go [hunk 3]
```diff
@@ -513,6 +514,48 @@ func adminResetMfa(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	newPassword := props["new_password"]
+	if len(newPassword) < model.MIN_PASSWORD_LENGTH {
+		c.SetInvalidParam("adminResetPassword", "new_password")
+		return
+	}
+
+	var user *model.User
```
Comment (crspeller) [inline api/admin.go]: From here on to line 555 is common to both. Helper function?

Comment (jwilander) [inline api/admin.go]: Yup, good point

**Comment by jwilander** [comment]
Ready for review

**Comment by crspeller** [comment]
+1

**Pull request merged**